harness/determined/layers/_worker_process.py

```diff
@@ -245,111 +195,43 @@ def _launch_python_subprocess(self) -> subprocess.Popen:
         ]
         return subprocess.Popen(python_cmd)
 
-    def _do_startup_message_sequence(self) -> None:
-        # Wait for a ConnectedMessage from every worker.
-        responses, exception_received = self.broadcast_server.gather_with_polling(
-            self._health_check
-        )
-
-        if exception_received:
-            raise det.errors.WorkerError("Training process died.")
-
-        for response in responses:
-            check.is_instance(
-                response,
-                ipc.ConnectedMessage,
-                f"Did not receive ConnectedMessage from worker. Got: {response}",
-            )
-            response = cast(ipc.ConnectedMessage, response)
-            self._worker_process_ids.append(response.process_id)
-
-        # Avoid hangs when sending to the broadcast client.
-        self.broadcast_server.safe_start(self._health_check)
-
-    def run(self) -> None:
-        """
-        The main control loop for controlling worker processes.
-        """
-
-        try:
-            self._do_startup_message_sequence()
-            for wkld, args, response_func in self.workloads:
-                response_func(self._send_recv_workload(wkld, args))
-        finally:
-            self.broadcast_server.close()
-
-            # If we wait for delete to be called when the interpreter shutdowns, we sometimes get a
-            # (harmless) stack trace from delete a socket with a `None` value. This is due to race
-            # conditions between when the interpreter deletes the weakref module and when pyzmq
-            # calls the weakref module in a __del__ method. We work around this by triggering the
-            # garbage collection earlier.
-            del self.broadcast_server
+    def run(self) -> int:
+        with self.pid_server:
+            try:
+                self.pid_server.run(self._health_check)
+            except det.errors.WorkerError:
```

Comment:
this looks like a case where context manager style is too much.
- start is done explicitly in `__init__` anyway, so `PIDServer#__enter__` here is noop.
- `PIDServer#run` could just close itself on return (or on errors).

this'll cut a level of indentation here without affecting semantics.

Reply:
have another look.  I don't disagree with your comment but since I rewrote the launch layer I'm not sure it's still relevant.